AI/VCAI/HeroManager.cpp

```diff
@@ -0,0 +1,237 @@
+/*
+* HeroManager.cpp, part of VCMI engine
+*
+* Authors: listed in file AUTHORS in main folder
+*
+* License: GNU General Public License v2.0 or later
+* Full text of license available in license.txt file, in main folder
+*
+*/
+
+#include "StdInc.h"
+#include "HeroManager.h"
+
+#include "../../CCallback.h"
+#include "../../lib/mapObjects/MapObjects.h"
+#include "../../lib/CHeroHandler.h"
+
+SecondarySkillEvaluator HeroManager::wariorSkillsScores = SecondarySkillEvaluator(
+	{
+		std::make_shared<SecondarySkillScoreMap>(
+			std::map<SecondarySkill, float>
+			{
+				{SecondarySkill::DIPLOMACY, 2},
+				{SecondarySkill::LOGISTICS, 2},
+				{SecondarySkill::EARTH_MAGIC, 2},
+				{SecondarySkill::ARMORER, 2},
+				{SecondarySkill::OFFENCE, 2},
+				{SecondarySkill::AIR_MAGIC, 1},
+				{SecondarySkill::WISDOM, 1},
+				{SecondarySkill::LEADERSHIP, 1},
+				{SecondarySkill::INTELLIGENCE, 1},
+				{SecondarySkill::RESISTANCE, 1},
+				{SecondarySkill::MYSTICISM, -1},
+				{SecondarySkill::SORCERY, -1},
+				{SecondarySkill::ESTATES, -1},
+				{SecondarySkill::FIRST_AID, -1},
+				{SecondarySkill::LEARNING, -1},
+				{SecondarySkill::SCHOLAR, -1},
+				{SecondarySkill::EAGLE_EYE, -1},
+				{SecondarySkill::NAVIGATION, -1}
+			}),
+		std::make_shared<ExistingSkillRule>(),
+		std::make_shared<WisdomRule>(),
+		std::make_shared<AtLeastOneMagicRule>()
+	});
+
+SecondarySkillEvaluator HeroManager::scountSkillsScores = SecondarySkillEvaluator(
+	{
+		std::make_shared<SecondarySkillScoreMap>(
+			std::map<SecondarySkill, float>
+			{
+				{SecondarySkill::LOGISTICS, 2},
+				{SecondarySkill::ESTATES, 2},
+				{SecondarySkill::PATHFINDING, 1},
+				{SecondarySkill::SCHOLAR, 1}
+			}),
+		std::make_shared<ExistingSkillRule>()
+	});
+
+void HeroManager::init(CPlayerSpecificInfoCallback * CB)
+{
+	cb = CB;
+}
+
+void HeroManager::setAI(VCAI * AI)
+{
+	ai = AI;
+}
+
+float HeroManager::evaluateSpeciality(const CGHeroInstance * hero) const
+{
+	auto heroSpecial = Selector::source(Bonus::HERO_SPECIAL, hero->type->ID.getNum());
+	auto secondarySkillBonus = Selector::type(Bonus::SECONDARY_SKILL_PREMY);
+	auto specialSecondarySkillBonuses = hero->getBonuses(heroSpecial.And(secondarySkillBonus));
+	float specialityScore = 0.0f;
+
+	for(auto bonus : *specialSecondarySkillBonuses)
+	{
+		SecondarySkill bonusSkill = SecondarySkill(bonus->subtype);
+		float bonusScore = wariorSkillsScores.evaluateSecSkill(hero, bonusSkill);
+
+		if(bonusScore > 0)
+			specialityScore += bonusScore * bonusScore * bonusScore;
+	}
+
+	return specialityScore;
+}
+
+float HeroManager::evaluateFightingStrength(const CGHeroInstance * hero) const
+{
+	return evaluateSpeciality(hero) + wariorSkillsScores.evaluateSecSkills(hero) + hero->level * 1.5f;
+}
+
+std::map<HeroPtr, HeroRole> HeroManager::getHeroRoles() const
```

Comment:
This logic should be unified with VCAI::primaryHero()

Reply:
Maybe. But this is a very big and complex question which I am not much interested in.

Reply:
The main difference between primary hero and this logic is that army value is not taken into account because the army can be given to any hero making it primary which does not look correct.